### ISSUE TITLE
tracing: include tracepoints in GUIX builds

### ISSUE THREAD

**Issue by 0xB10C** [bitcoin/bitcoin#23297]
There has been light conceptual agreement on including the USDT based tracepoints in Bitcoin Core release builds. This, for example, enables user to hook into production deployments, if they need to. Binaries don't have to be switched out. This is possible because we don't do [expensive computations](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#no-expensive-computations-for-tracepoints) only for the tracepoints and the tracepoints are NOPs when not used.

There is a slight chance that the GUIX build on the current master already includes the tracepoints. I have not done an GUIX build on a branch with the tracepoints merged yet. This can be tested on a `bitcoind` binary using one of the methods mentioned in the [tracing documentation section "Listing avaliable tracepoints"](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#listing-available-tracepoints).

If not present, making the `systemtap` headers (`sys/sdt.h`) available during the GUIX build should build `bitcoind` with the tracepoints.

**Comment by MooCricket** [comment]
แบบไหนดีครับ

**Comment by MooCricket** [comment]
ไก้ไหมครับ

**Comment by 0xB10C** [comment]
> There is a slight chance that the GUIX build on the current master already includes the tracepoints. I have not done an GUIX build on a branch with the tracepoints merged yet. This can be tested on a `bitcoind` binary using one of the methods mentioned in the [tracing documentation section "Listing avaliable tracepoints"](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#listing-available-tracepoints).

A GUIX build on 76392b042e98 didn't include the tracepoints. The systemtap headers need to be added.

**Comment by 0xB10C** [comment]
I've been exploring this a bit more. Currently two approaches, non of them really great. 

- a) Package the systemtap `sys/sdt.h` header as GUIX package in [manifest.scm](https://github.com/bitcoin/bitcoin/blob/master/contrib/guix/manifest.scm) and make it available as _include_ during configure and build.
  - See GUIX package definition below.
  - While the package is available during configure and build, the header includes not are picked up automatically. Likely, an additional include needs to be set. 
- b) Include systemtap as package in `depends`
  - We don't need systemtap. We only need the `sys/sdt.h` header file.
  - depends are usually compiled, we don't want and need to compile systemtap here.

```scheme
(define-public sys-std-h
  (package
    (name "sys-sdt-h")
    (version "4.6")
    (source (origin
              (method url-fetch)
              (uri (string-append "https://sourceware.org/systemtap/ftp/releases/systemtap-" version
                                  ".tar.gz"))
              (sha256
               (base32
                "1bps2926x9m8mp2684wwwwjm4yb12mpylbnvnffk889b4c4p7yw0"))))
    (build-system copy-build-system)
    (arguments
      '(#:install-plan
        '(("includes/sys/sdt.h" "include/sys/sdt.h"))))
    (synopsis "Systemtap header files defining tracing probe marcos")
    (description "Systemtap sys/sdt.h header file for Userspace, Statically Defined Tracing support.")
    (home-page "https://sourceware.org/systemtap/")
    (license license:gpl2)))
```

**Comment by 0xB10C** [comment]
Seems to be possible with option b) too. The configure/build/postprocess steps in the depends are optional and normal shell commands can be used. I haven't figured out where to copy the `includes/sys/sdt.h` file from systemtap so that it makes it into `depends/x86_64-linux-gnu/include/`. 

I think option b) is cleaner than option a).

**Comment by 0xB10C** [comment]
@fanquake helped getting the includes copied. In this case it was: (note the `&& \` as only the first command has the correct ENV var set)

``` Makefile
define $(package)_preprocess_cmds
  mkdir -p $($(package)_staging_prefix_dir)/include/sys && \
  cp -r includes/sys $($(package)_staging_prefix_dir)/include
endef
```

He also commented:

> One catch is that if we don't run configure and make, then we don't end up with a properly named and filled out sdt-config.h. That is included from sdt.h, and determines whether the probes can use certain assembly, see _SDT_ASM_AUTOGROUP.

> Maybe this is mostly a legacy check, and we can assume that we can use it, although then we'd still have to modify the sdt.h header to drop the include and make sure _SDT_ASM_SECTION_AUTOGROUP_SUPPORT is defined ourselves.

> The annoying thing about having to run configure and make is that we'll need to install a bunch of dependencies.

The relevant section from sdt.h is

``` h
/* If the assembler supports the necessary feature, then we can play
 nice with code in COMDAT sections, which comes up in C++ code.
 Without that assembler support, some combinations of probe placements
 in certain kinds of C++ code may produce link-time errors.  */
#include "sdt-config.h"
#if _SDT_ASM_SECTION_AUTOGROUP_SUPPORT
# define _SDT_ASM_AUTOGROUP "?"
#else
# define _SDT_ASM_AUTOGROUP ""
#endif
```

As we don't need Systemtap itself, running configure and make would require additional dependencies and slow the depends build down. Both, the include of `sdt-config.h` in `sdt.h` and the `sdt-config.h.in` file were added and last touched in 2010 ( 
[blame sdt.h](https://sourceware.org/git/?p=systemtap.git;a=blame;f=includes/sys/sdt.h;h=48cb7263f325ff8094b4288249a24b65dd00ab84;hb=HEAD), [blame sdt-config.h.in](https://sourceware.org/git/?p=systemtap.git;a=blame;f=includes/sys/sdt-config.h.in;h=7a3373cca0d9d52e74c9096a9db48cc37d897c5e;hb=HEAD)), so it's likely that the feature is supported by now and it should be possible to drop the include in a patch.

**Comment by 0xB10C** [comment]
gcc seems to have a problem with Systemtap 4.6 (Nov 2021). I'll be using Systemtap 4.5 (May 2021) for now. From what I can tell, this is a know problem and it should be fixed in 4.7:

https://sourceware.org/git/?p=systemtap.git;a=commit;h=1d3653936fc1fd13135a723a27e6c7e959793ad0
https://bugzilla.redhat.com/show_bug.cgi?id=2028798

**Comment by 0xB10C** [comment]
I've successfully ran a GUIX build with the new `sys/sdt.h` systemtap depends as described above. All host triples with linux in the name contain the tracepoints (checked with `readelf -n bitcoind`). I haven't looked at the macOS and Windows binaries. Maybe the macOS binaries contain tracepoints that can be listed with e.g. `dtrace` on macOS?

|      host-triplet     	| contains tracepoints 	| runs without crashing 	| tracepoints usable 	|
|:---------------------:	|:--------------------:	|:---------------------:	|:------------------:	|
|   aarch64-linux-gnu   	|  :heavy_check_mark:  	|                       	|                    	|
|  arm-linux-gnueabihf  	|  :heavy_check_mark:  	|                       	|                    	|
|  powerpc64-linux-gnu  	|  :heavy_check_mark:  	|                       	|                    	|
| powerpc64le-linux-gnu 	|  :heavy_check_mark:  	|                       	|                    	|
|   riscv64-linux-gnu   	|  :heavy_check_mark:  	|                       	|                    	|
|    x86_64-linux-gnu   	|  :heavy_check_mark:  	|                       	|                    	|
| x86_64-apple-darwin19 	|    :grey_question:   	|                       	|                    	|
|   x86_64-w64-mingw32  	|    :grey_question:   	|                       	|                    	|

Will do a bit of basic testing and then open a PR.
